john-royal/csc-456-paw
# Jobanpreet Singh Session 6 - Assignment: Unit/Integration/E2E tests and GitHub Actions on Pull request

Changes to code:

Unit Test:
Located under the directory: `tests/unit/PostCard.test.tsx` 

The [PostCard.test.tsx](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/tests/unit/PostCard.test.tsx) is a unit test for the [PostCard.tsx](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/src/components/PostCard.tsx) component. This unit test aims to verify the functionality of the PostCard component in a React application. The first test ensures that the PostCard component properly renders with provided post data, including elements like user profile photo, username, pet image, like count, and comment count. The second test verifies that the PostCard component gracefully handles invalid post data by throwing an error message when essential fields are missing, such as username, pet profile photo, pet image, like count, and comment count.

Integration Test:
Located under the directory: `tests/integration/Contact.test.tsx`

The [Contact.test.tsx](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/tests/integration/Contact.test.tsx) is an integration test for [Contact.tsx](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/src/routes/Contact.tsx) page. This integration test evaluates the behavior of the Contact form component in a React application, considering interactions with Firebase Firestore. In the first test, it simulates user input for name, email, and message fields, then triggers the form submission by clicking the 'Send Message' button. Finally, it confirms the display of a success message ('Message Sent!') which is only shown when Firebase returns a success promise. The second test mocks an error scenario when submitting to Firebase and checks if the appropriate error message is handled. Empty values are entered into all fields, triggering the form submission. It verifies that no success message is displayed due to the error scenario.

End to End Test:
Located under the directory: `tests/endtoend/SignIn.spec.ts`

The [SignIn.spec.ts](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/tests/endtoend/SignIn.spec.ts) is an integration test for [auth.sign-in.tsx](https://github.com/programmerjoban/csc-456-paw/blob/0a24d8ecf41591afae1b34c9276c87ab89ae6fda/src/routes/auth.sign-in.tsx) component. This end-to-end test, utilizing Playwright, focuses on the sign-in process of a web application. It mocks the sign in behavior by filling in email, password and clicking the signin button. On Success, it verifies that the user is redirected to the root path by checking the current URL against the expected root URL.

Code coverage
Description:
All of the tests for the PostCard component, Contact form, and Authentication end to end test are passed. The results are pushed to github but can also we seen when the github actions run for “Executing Unit and Integration Test, and Generating code coverage report”

| File                 | % Stmts | % Branch | % Funcs | % Lines |   
| ---------------     | -------      | --------       | -------      | -------      | 
| All files            | 72.34     | 83.33        | 38.88     | 78.04     |                      
| components    | 56          | 71.42        | 10          | 66.66     |                      
| PostCard.tsx   | 56          | 71.42        | 10          | 66.66     | 
| routes              | 90.9       | 100           | 75          | 90          |                      
| Contact.tsx      | 90.9       | 100           | 75          | 90          |  

There is significant progress in code coverage after writing the tests as compared to before the tests . The results show that overall 72% of the statements are tested in all files. The unit test for the PostCard component is covering 56% of the code statements and 66% of the lines. The integrations test for the Contact us page is covering 90% of the code statements and 90% of the lines. This makes sense since I wrote only 4 tests in total for the unit and integration. In the future, I will add more tests to the components so that 100% of the code is covered. The Playwright test were tested in the UI mode and the test worked successfully.

SetUp:

No additional setup needed, the CI should install all the required dependencies. However if you run this CI locally using “act” you have to pass a github access token.

Run this command to test GitHub Actions locally with act: 
`act --artifact-server-path /tmp/artifacts -s GITHUB_TOKEN="$(gh auth token)"`

Overall the following Github Actions are performed on pull request:
1. Check out PR
2. Setup node
3. Install Dependencies
4. Install Playwright
6. Executing Unit and Integration Test, and Generating code coverage report
7. Upload Code Coverage Test
8. Executing Playwright end to end Test
9. Upload Playwright test
